### DISCUSSION TITLE
[Fooocus 2.1.822] Fooocus Inpaint or Outpaint (Midjourney Left/Right/Top/Bottom)

### DISCUSSION THREAD

**Discussion by lllyasviel** [lllyasviel/Fooocus#414] [Announcements]
(**Last Update, 2023 Nov 19**)

Fooocus 2.1.822 added inpaint and outpaint similar to Midjourney’s “left/right/top/bottom” arrows:

![image](https://github.com/lllyasviel/misc/assets/19834515/aec79f55-ad33-4537-a162-3412d13749e0)

In Fooocus it looks like this:

![image](https://github.com/lllyasviel/misc/assets/19834515/f97b0068-1544-4eac-8179-44c12f7002f4)

![image](https://github.com/lllyasviel/misc/assets/19834515/a391d088-53e3-4d0d-878c-34fc8c74a69c)

This “Fooocus Inpaint” is not only a user-friendly UI but also an algorithm production. Fooocus uses its own algorithm DPMPP Fooocus inpaint, and also uses Fooocus's own control model to minimized the influence to the style of base models. The method is partially inspired by [Diffusion-based semantic image editing with mask guidance](https://openreview.net/forum?id=3lge0p5o-M-).

In the entire open source community of SDXL, the Fooocus is the only software that allows you to use control-model-based inpaint with arbitary base models.

Note that this method, in general, expects processing generated images with unchanged or minorly changed prompts. We encourage users to drag image like this:

![image](https://github.com/lllyasviel/misc/assets/19834515/51696d14-676e-48d3-bce1-5d6156307f3b)

# Inpaint Example

![image](https://github.com/lllyasviel/misc/assets/19834515/57a4282f-c2db-44ea-b951-61c39a041efa)

![image](https://github.com/lllyasviel/misc/assets/19834515/aad4d6f7-c9c5-47c5-8dc6-7cede1165310)

# Outpaint Example

Use the pan to outpaint:

![image](https://github.com/lllyasviel/misc/assets/19834515/727918ef-3a85-4a3d-8d45-fb570adacb9d)

![image](https://github.com/lllyasviel/misc/assets/19834515/30205039-a8bf-4f3a-8791-d64603252957)

<a name="detail"></a>

# Advanced: Detail Fix (face, hand, eyes, ...)

From time to time, you may get images like this:

![image](https://github.com/lllyasviel/misc/assets/19834515/674558bb-4a8f-48a1-9a9a-a6dfc09d1b14)

Everything is perfect, but the eyes / faces are just annoying:

![image](https://github.com/lllyasviel/misc/assets/19834515/ce1dbaf7-eebc-4892-b0cc-a724c71ac7f4)

Now you can fix details like this:

![image](https://github.com/lllyasviel/misc/assets/19834515/a43f7787-eac0-4d4c-aafb-aaa99554cc14)

![image](https://github.com/lllyasviel/misc/assets/19834515/efe51c21-5e91-4b28-bc9b-98da80588a56)

Note that this interface also allows you to add some prompts just for this inpaint pass. If you are too lazy, you can also use the shortcuts here:

![image](https://github.com/lllyasviel/misc/assets/19834515/aefd606e-4d71-420f-8ac7-5eb25f6dcab4)

and you get result like these

![image](https://github.com/lllyasviel/misc/assets/19834515/15cc6625-f58b-4727-ab9b-b54f4c22f70a)

![image](https://github.com/lllyasviel/Fooocus/assets/19834515/c86d5da9-51f0-4bd7-a36f-9cb6fc14f9c9)

Another example:

before:

![image](https://github.com/lllyasviel/misc/assets/19834515/1be0d01e-d9d1-4714-9635-728c429a89b5)

after:

![image](https://github.com/lllyasviel/misc/assets/19834515/37c98944-59fd-47f0-aa53-4eea2559e4af)

# Advanced: Add New Objects / Change Background

You can always add new objects just using default setting. But if the modification is too large, a better idea is to use 

![image](https://github.com/lllyasviel/misc/assets/19834515/fa5dcac3-8712-4fc6-9f96-5cc5ab2afb11)

For example:

![image](https://github.com/lllyasviel/misc/assets/19834515/ea1f8bb1-4e81-406a-882b-75d51ff40a2c)

Change Background:

This "modification" mode will completely ignore original content and is good at handling images with solid color backgrounds like

![image](https://github.com/lllyasviel/misc/assets/19834515/6144530c-65fe-4c4e-aa82-a495bd0ac1a7)

![image](https://github.com/lllyasviel/misc/assets/19834515/ef8d8b78-cac5-4422-8a4e-90fbc11eaaa5)

# Very Advanced: Multiple-pass inpaint

You can first use general inpaint to get the image, drag the image to inpaint input again, and then change the mode to "Improve Detail" and then inpaint again. In this way you will get extremely high quality result. (Improve Detail can be used to improve details of anything, not only faces or hands).

**Comment by AssassinFlash** [comment]
hello, There's a trouble when I'm using the inpaint. I dragged a photo with a house, masked the wall and typed "battery storage on the wall". Then I clicked generate, it just stuck and the cmd just said image processing but without further more.

**Reply by NovaAI77** [reply]
hit the space bar in the cmd window

**Comment by MoonRide303** [comment]
@lllyasviel Problem is that base SDXL model wasn't trained for inpainting / outpainting - it delivers far worse results than dedicated inpainting models we've had for SD 1.x / 2.x. IMO we should wait for availability of SDXL model trained for inpainting before pushing features like that. You can see discussion on this subject and comparison of the results in [diffusers issue 4392](https://github.com/huggingface/diffusers/issues/4392#issuecomment-1664248008).

Same test case in Fooocus v2.0.60 below.

Input image:
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/93185c37-bf1e-4633-8a27-25e5112c1838)

Original mask:
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/0d967981-600c-4289-b3ed-944c971e7bf8)

Manually applied mask in Fooocus:
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/e1b96416-0dac-4f92-960f-56c606309eb8)

Output:
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/924df007-faec-4a93-b098-cdb81236395c)

**Reply by lllyasviel** [reply]
"DPMPP Fooocus Seamless 2M SDE Karras Inpaint Sampler" is designed for generated images with unchanged or minor edited prompts, similar to MJ use cases

**Reply by lllyasviel** [reply]
resolved in 2.0.65

**Reply by MoonRide303** [reply]
@lllyasviel I did some tests on v2.0.65 - it's noticeably better than v2.0.60:
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/c7378b7d-e05c-42e9-87f3-76470ee33b57)

But it's still lagging behind old inpainting models - output from SD 1.5 inpainting for comparison:
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/b7eaf0c0-7870-4a5b-b90d-478e3b88b948)

Made with diffusers:
```
import torch
from diffusers import StableDiffusionInpaintPipeline
from diffusers.utils import load_image

pipe = StableDiffusionInpaintPipeline.from_pretrained(
    "runwayml/stable-diffusion-inpainting", torch_dtype=torch.float16, variant="fp16"
)
pipe.to("cuda")

img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"

init_image = load_image(img_url).convert("RGB")
mask_image = load_image(mask_url).convert("RGB")

prompt = "an empty bench"
generator = torch.Generator()

generator.manual_seed(0)
image = pipe(prompt=prompt, image=init_image, mask_image=mask_image, width=1024, height=1024, generator=generator).images[0]
image.save("inpainting-test-sd15-x2.png")

```

**Reply by MoonRide303** [reply]
I tried with a bit bigger masks (single area) and seems to work better, then:
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/c46162f2-8045-425c-a41c-ca12ae60d8e1)

It looks like it confused scale at some point, though - the inpainted pattern on the bench looks about twice as big as it should be.

**Reply by lllyasviel** [reply]
have several other models but will save some for Fooocus V3
also some inpaint models are only better before offset merge, some others (especially external models) are better after changing base

**Comment by onlinedear** [comment]
cool

**Reply by Wontfallo** [reply]
@lllyasviel  Fooocus V3 you say??  (❁´◡`❁)

**Comment by holyvslin** [comment]
Fantasy!
With selected area, Generate again with same prompt and change image number to 1.
New "real" image without selected area will be output.

**Comment by lllyasviel** [comment]
Update 2.0.65:

Fooocus now uses its own inpaint model for inpaint/outpaint. The model is 1.2GB and it will be downloaded in the first time you inpaint an image.

![image](https://github.com/lllyasviel/Fooocus/assets/19834515/05d33e03-691f-4799-9b27-46eb0e79594f)

![image](https://github.com/lllyasviel/Fooocus/assets/19834515/9e424f9c-62fa-44ed-8d50-fe3d5c1835c1)

![image](https://github.com/lllyasviel/Fooocus/assets/19834515/b295408d-869e-4865-8660-5d999103749c)

**Reply by 12-365** [reply]
how ca do import my back ground and put my secretary

**Comment by bluntstone** [comment]
new version is broken on linux, complains cuda version is wrong. Going to try to install on windows.

**Reply by lllyasviel** [reply]
please follow official installation guide on linux

**Reply by bluntstone** [reply]
I did, on the fresh install its broken, and the one that was working I did a git pull and it broke. It was working no issues about a week ago.  I will try again and post the error. I deleted the install so I can't post the error now.

**Reply by bluntstone** [reply]
Working now, I think I didn't do a step. the install instructions maybe changed. Not really sure, I am using a fresh ubuntu server install and its all working. I don''t think it did "pip install pygit2==1.12.2" the last time.

**Comment by MoonRide303** [comment]
@lllyasviel I've merged changes from v2.0.76 into [MRE testing branch](https://github.com/MoonRide303/Fooocus-MRE/tree/testing) (using current ComfyUI as backend), but I am observing color problems in inpainting and outpainting modes, like this:
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/9556e657-506e-4ae2-9c84-e227c5c5539b)

Any idea what might be causing that reddish tint? I tried to keep the data processing as in vanilla, and normal generation works fine.

PS I get same reddish tint over whole image during inapainting preview:
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/e19f19f0-51d1-49ad-9a51-aae32355c4d3)

When I've enabled saving input images in InpaintWorker self.image_raw seems to be fine:
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/1bc0f8c1-faf5-4bd8-bf3e-ac377ed12023)

**Reply by lllyasviel** [reply]
is image 0-255 range?

**Reply by MoonRide303** [reply]
@lllyasviel It's the same test image as before, resized inpainting sample provided originally by CompVis:
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/e6518cca-2aa3-4198-be7b-d933d51ed6ae)

Channels histograms:
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/7d306d4d-25d8-4caa-8904-82bcf2e7536f)
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/9f61b880-c21f-49ea-8ce2-45a212bf786a)
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/1e78d0a6-bb9c-435b-93d8-42e341eb7e7b)

It works fine in vanilla v2.0.76, but I don't see why inpainting behaves differently in MRE :/.

**Reply by lllyasviel** [reply]
do you have other examples or is it always more red and no other colors

**Reply by MoonRide303** [reply]
@lllyasviel I got it - in case of MRE default sampler_name ('dpmpp_2m_sde_gpu') wasn't handling inpainting colors properly. When I set sampler_name to 'dpmpp_fooocus_2m_sde_inpaint_seamless' for inpainting / outpainting workflows it looks good, now.

It would be nice to implement this inpainting-specific parts in more generic way, so it would work with any base sampler. Wouldn't proposing pull requests for ComfyUI (so we wouldn't need all that patching on the Fooocus side) be better long-term solution?

**Reply by lllyasviel** [reply]
cool

**Comment by fatinghenji** [comment]
Hello, I have not come into contact with Midjourney before, so the "arrow" mentioned in the above post confused me. Can you give a more detailed explanation?

**Reply by OperationTwelve** [reply]
those arrows when selected will outpaint in those directions.

**Comment by sirilucas** [comment]
Hello, I'm having trouble with inpainting. Please note that my coding skills are almost zero.

At the moment I'm trying to inpaint some areas masked in a previusly generated image, it seems that the code runs nominal, but at the very bottom it stops at a RuntimeError that I can't solve:

```
File "D:\Descargas\SETUP\Fooocus_win64_2-0-50\python_embeded\lib\site-packages\torch\serialization.py", line 283, in __init__
    super().__init__(torch._C.PyTorchFileReader(name_or_buffer))
RuntimeError: PytorchStreamReader failed reading zip archive: failed finding central directory
```

I hope some of you can help me! Thanks in advance.

**Comment by tmprabubiz** [comment]
Dear lllyasviel

Inpainting, Outpainting and Upscaling is not working for me

I am just getting error messge "connection errored out" and in the command box the number "1006" appears

Why the folders like "embeddinggs', "inpaint", "style_,models" and others are empty

I am getting following error message while loading
///missing {'cond_stage_model.clip_g.transformer.text_model.embeddings.position_ids'}////
///missing {'cond_stage_model.clip_g.transformer.text_model.embeddings.position_ids'}///


Should I install any models in the above empty folders to make Outpainting and Upscaling work


Please guide

**Reply by barepixels** [reply]
Do you have these 5 files?

E:\Fooocus-lllyasviel\Fooocus\models\inpaint\inpaint.fooocus.patch
E:\Fooocus-lllyasviel\Fooocus\models\inpaint\fooocus_inpaint_head.pth
E:\Fooocus-lllyasviel\Fooocus\models\upscale_models\fooocus_upscaler_s409985e5.bin
E:\Fooocus-lllyasviel\Fooocus\models\vae_approx\xlvaeapp.pth
E:\Fooocus-lllyasviel\Fooocus\models\vae_approx\taesd_decoder.pth

**Reply by tmprabubiz** [reply]
""E:\Fooocus-lllyasviel\Fooocus\models\inpaint\inpaint.fooocus.patch""
Above file is missing
How to get it. 
I have made three fresh installations already. But I could not get rid of the problem.
Please guide

**Reply by MoonRide303** [reply]
@tmprabubiz Those supplementary models should be automatically downloaded from these locations:
-  https://huggingface.co/lllyasviel/fooocus_inpaint/tree/main
- https://huggingface.co/lllyasviel/misc/tree/main

You didn't have any errors in the console about failed downloads?

**Reply by barepixels** [reply]
MoonRide303 can you provide a tree lists all the files or directories.  Now am paranoid if I miss anything haha

One thing I notice, my Fooocus-MRE have "clip_vision_g.safetensors" but Fooocus-lllyasviel & Fooocus-Runew0lf does not

**Reply by barepixels** [reply]
What is this https://huggingface.co/lllyasviel/misc/blob/main/fooocus_expansion.bin

Both Fooocus-MRE and Fooocus-lllyasviel have "pytorch_model.bin"  in E:\Fooocus-lllyasviel\Fooocus\models\prompt_expansion\fooocus_expansion

**Reply by captainpd** [reply]
This is a GPT-2 NLP model for extanding your current prompts for better performance. It will change your prompts and print out the result on the cmd.

**Reply by MoonRide303** [reply]
@barepixels And clip_vision_g is used for Revision workflow, only - it's not needed in vanilla.

**Reply by barepixels** [reply]
So [fooocus_expansion.bin](https://huggingface.co/lllyasviel/misc/blob/main/fooocus_expansion.bin) is the same as **pytorch_model.bin**  BUT both Fooocus-MRE and Fooocus-lllyasviel prefer **pytorch_model.bin** ?  a file with a generic name?  lol

**Comment by barepixels** [comment]
This video explains why Fooocus' Outpainting area is blurry
https://youtu.be/VJymbJ97WDQ?si=TYNKhARgmvKvOEi2

**Reply by MoonRide303** [reply]
@barepixels That's kinda expected - to get high quality output you shouldn't inpaint / outpaint region bigger than 1024x1024 at a time (at least with current inpainting implementation). That's also the reason why you get better results by separately outpainting left, THEN separately right, etc.

Just please mind doing inpainting / outpainting as tiles will side effects:
1) there might be artifacts at tiles edges (as explained [around 7:00 in that video](https://youtu.be/VJymbJ97WDQ?si=mmsr25IAouuhtDhK&t=419)),
2) process will take more time.

**Comment by AsenAsenAsen** [comment]
when the picture width exceeds the maximum width of the preview box,cannot apply a mask that exceeds the part,
![image](https://github.com/lllyasviel/Fooocus/assets/147331264/9e0b2d88-2b63-4c09-aace-fa994e895025)

**Reply by lllyasviel** [reply]
gradio bug
uncheck advanced to make it wider

**Comment by lllyasviel** [comment]
[Important] Fooocus Inpaint V2 Beta Test: https://github.com/lllyasviel/Fooocus/discussions/615

**Comment by funkysugar1983** [comment]
![Snipaste_2023-11-10_14-51-06](https://github.com/lllyasviel/Fooocus/assets/149927943/74182678-3047-4f3f-bab4-24600b9e9b9b)
can somebody tell me why this happened? can I download the model manually and where should I put the files? thx in advance.

**Reply by ycyy** [reply]
[https://huggingface.co/lllyasviel/fooocus_inpaint/tree/main]() download from this link. put into `Fooocus\models\inpaint` this path.

**Reply by funkysugar1983** [reply]
thks so much. it's working

**Comment by lllyasviel** [comment]
2023 Nov 19:
beta test ends and this post is updated.

**Comment by lanyusan** [comment]
@lllyasviel 

Just noticed the new advanced inpainting options and tested the add new objects/change background.

It worked great.

Thanks a lot for this great enhancement.

![image](https://github.com/lllyasviel/Fooocus/assets/56706512/a8b96a6e-582d-4ace-bbdd-bce8b0b27956)

**Reply by intangir** [reply]
what is that region you defined there? and how did you define it? not with the normal mask painting brush by the look of it?

**Comment by 3Diva** [comment]
The new "Advanced: Detail Fix (face, hand, eyes, ...)" and "Advanced: Add New Objects / Change Background" is OUTSTANDING! Thank you so much for the amazing update and new features! This is VERY helpful stuff! AWESOME!

**Comment by SergeyArtgraver** [comment]
Есть ли возможность в фокусе восстановления или ретуширования обычных фотографий?
Спасибо, если будет ответ.

**Comment by Cyoshi** [comment]
In background modification mode etc. Please add an eraser to erase an area. Right now you can either cancel the last step or erase all of it. and sometimes you can select more than you need and have to start over again

**Reply by 3Diva** [reply]
Yeah, getting precision masking is a little difficult without the ability to erase. But coding something like that might be pretty tough.

**Reply by Cyoshi** [reply]
I hadn't thought about the complexity of realizing this feature. I'm evaluating the experience as a designer
if it ever comes to fruition, a couple more handy features
It would be very convenient if you could draw straight lines (vertical or horizontal) with Shift 
and draw a line from the previous click to the current one (as it is implemented in Photoshop) when clicking with shift

The product is already awesome, but in case they think about what feature to add in the future and remember this comment.

**Comment by williamsblood** [comment]
Thank you for inpainting @lllyasviel , I never thought I could make this kind of photos using  SDXL!
![image (15)](https://github.com/lllyasviel/Fooocus/assets/144069983/9299383c-6d87-4a40-83a7-472eee515249)

![image (6)](https://github.com/lllyasviel/Fooocus/assets/144069983/63a1c670-2b69-460a-a08b-0cb45e8c4aa4)

**Comment by jclevet** [comment]
Could we influence the inpainting with a image prompte like PyraCanny or CPDS. The two tabs are mutually exclusive, right?

**Reply by 3Diva** [reply]
@jclevet I believe you can use image prompts to influence inpainting in Developer Debug Mode. Go to the Advanced tab, click on Developer Debug Mode, go to the Control tab, then click on "Mixing Image Prompt and Inpaint". 

I've used it a few times and it works ok with Method: "Inpaint or Outpaint (default)", but seems to work the best (IMO) with "Modify Content". Keep in mind that it doesn't seem to work with "Improve Detail" unless you're using "FaceSwap".

**Reply by jclevet** [reply]
Wow ! amazing ! Thanks a lot. Fooocus is incredible.

**Reply by 3Diva** [reply]
@jclevet Right? It's awesome! I used to use A1111 but since lllyasviel has implemented such great features into Fooocus I no longer need to use AUTOMATIC1111. Which is great because the memory leak of A1111 really hit my computer hard and made using it a pain and a real chore. Fooocus and it's speed + amazing image quality makes it superior, imo, to pretty much every other text-to-image generator out there.

**Reply by jclevet** [reply]
like you: I have invokeAI, A1111 and comfyUI installed. I only use Fooocus anymore.

**Comment by Don-Rez** [comment]
how do u fix the hands of the models, they either look unnagurall or 7 fingers

**Comment by artistry-intelligence-solutions** [comment]
is there any way to change the background to a specific image?

**Reply by 3Diva** [reply]
@artistry-intelligence-solutions You can try inpainting it in by linking Inpainting with Image Prompt. To enable a link between Image Prompt and Inpainting go to: Advanced >> Developer Debug Mode >> Mixing Image Prompt and Inpaint

**Reply by artistry-intelligence-solutions** [reply]
oh my goodness, you are such a star thank you @3Diva. Take care

**Comment by aimarxjg** [comment]
@MoonRide303 NEED HELP， I have obvious seam marks after expanding the picture, how should I avoid them (Use the brush tool to paint the white seam and create the picture again with the seam) I want to keep the vehicle and do not want to use the (variation - subtle) function to affect the car body paint。
![202829](https://github.com/lllyasviel/Fooocus/assets/132532636/c0e228bc-b4e0-4020-ac7d-ed8b9a917014)
![QQ图片20231208202836](https://github.com/lllyasviel/Fooocus/assets/132532636/aa8d48f6-e077-407e-b342-2f2041f9a998)

**Reply by 3Diva** [reply]
You might need to do a second pass of inpainting to clean up those lines. Have you tried that to see if it helps?

**Reply by aimarxjg** [reply]
Thank  reply, I have tried, the white brush has painted the area and there are obvious marks around, I will try again

**Comment by zeratyl888** [comment]
![Снимок экрана 2023-12-10 192229](https://github.com/lllyasviel/Fooocus/assets/37886027/d771be51-b3e7-43af-aba8-271952d4eb15)

Hello, I use Google resources through the cloud, tell me why I can't edit my image.

That is, to "improve the quality of the face" and so on.

I click "generate" it takes a few seconds and that's it, nothing, a blank screen.

**Comment by qicuili** [comment]
how can i use the api about Inpaint or Outpaint

**Comment by artistry-intelligence-solutions** [comment]
Hello Fooocus community. Hope everyone is having a safe, happy festive season. I'm writing a comic book, continuity is very important. So I rendrered my protagonist on a white background, and rendered his office. I'm trying to place the character in the office and hopefully get the renderer to render the correct angle of the office according to the action. It either gives me something completely different or a jumbled pixelated mess. I've tried unsuccessfully all the different options of _Image Prompt_ and _Input or Output Paint_.

![Screenshot 2023-12-19 123126](https://github.com/lllyasviel/Fooocus/assets/44393928/fa7939b8-6307-4bce-91a6-538e48832ddf)

![Screenshot 2023-12-19 123206](https://github.com/lllyasviel/Fooocus/assets/44393928/fb1b8e17-70ff-4283-ae3d-3c3ef29481e4)

![Screenshot 2023-12-19 123147](https://github.com/lllyasviel/Fooocus/assets/44393928/0b3ca210-f69a-4a67-93c0-d1ddd9f39b9e)

Obvsiously visual continuity is very important in any story telling so I'd really appreciate any and all help here. Thank you!

**Reply by simartem** [reply]
have a look at this : https://github.com/AIGCDesignGroup/ReplaceAnything

**Comment by jclevet** [comment]
Hi,
I don't know if you generated the child's image on Fooocus. But you could generate your subjects directly on a uniform green or blue background. This will allow you to embed them with keying tools on image software.
Add for exemple to your prompt : standing in front of a perfectly seamless vivid green screen, studio lighting.

I will work on 4 styles for subjects on green, blue, white and black backgrounds. I will offer them to Illyas if he is interested.

![green](https://github.com/lllyasviel/Fooocus/assets/9883751/015b3f33-77f9-4787-8afb-2e127b03d129)
![mask](https://github.com/lllyasviel/Fooocus/assets/9883751/d227a184-b337-4ca7-b5ad-2fd63d4dc756)

**Reply by artistry-intelligence-solutions** [reply]
> @jclevet thanks for getting back to me! Yes I did generate this with fooocus ^^

**Reply by siriume** [reply]
I hope that someday we can utilize `OpenPose` to generate motions or `mask` overlays for InPaint or OutPaint purposes.

**Comment by lawsonxwl** [comment]
Will the object removal model be released? Current inpaint model is not good at removing object.

**Comment by kpagcha** [comment]
I am also struggling to create a background for an input image with a white background.

**Comment by gutsalovpavel** [comment]
trying to generate clothes, by modifying content, but it is always generating green cloth... why green? is it the color of the mask for inpaint?

**Reply by gutsalovpavel** [reply]
ok, got it. it draws green objects when you have a pink background. strange anyway, but resolvable. we need to use grey background for proper colors

**Comment by pilotnikovk** [comment]
Hi, how can I insert an object into a picture, for example, I have a chair and an interior, and this chair needs to be inserted into my interior, is this possible?

**Reply by simartem** [reply]
may be you can have a look at this project : 

https://github.com/AIGCDesignGroup/ReplaceAnything

**Reply by 3Diva** [reply]
@pilotnikovk If you drop your image into the inpaint tab you should be able to paint over the area you want a chair to be and prompt for the chair to add it to the image.

**Comment by hyper-realist** [comment]
fooocus version: Google Colab public GPU server
Hi all, 
is anyone having trouble with an extended processing time? I feel while processing any of the image input features: Upscale or Variation/Image Prompt/Inpaint or Outpaint/Describe, never stops. 
Any thoughts to fix this?
 
![fooocuserror](https://github.com/lllyasviel/Fooocus/assets/156856177/b75c50f8-288c-44c3-b16e-99128dd56063)

**Reply by mashb1t** [reply]
@hyper-realist 
> Note that this Colab will disable refiner by default because Colab free's resources are relatively limited (and some "big" features like image prompt may cause free-tier Colab to disconnect). We make sure that basic text-to-image is always working on free-tier Colab.

Please check the terminal output and restart your Colab playbook execution if needed.

**Reply by hyper-realist** [reply]
Ahhhh,  Thanks for the quick reply, 
Does that mean I won't be able to use these features?

"Please check the terminal output and restart your Colab playbook execution if needed." what do you mean by this?

**Reply by mashb1t** [reply]
@hyper-realist 
> Does that mean I won't be able to use these features?

Most likely not, at least not reliably.
You can try to restart Colab on error or if it stopped and try again.

**Reply by hyper-realist** [reply]
I tried paying for the service yet it still seems to be doing the same thing

**Reply by mashb1t** [reply]
@hyper-realist please check if applying the proposed changes of https://github.com/lllyasviel/Fooocus/pull/1710 helps you to achieve your goal.

**Reply by hyper-realist** [reply]
Thank you for this, is this edited in the script? 
I'm unable to edit the script while running

**Comment by blackthornecl** [comment]
I have a problem when I draw a mask and set the message to "detailed girl face" instead of fixing the face, it makes me a whole new face of a detailed girl, any idea?

**Comment by calman13** [comment]
Hello. 
How can I combinate two images, for example a photo with my face with a photo generated by AI or vice versa. Also, is it possible to combine two models for example of civitAI, one anime and one realistic?

**Comment by Rotciv007** [comment]
Hello. I want to share an idea to add a feature to the Focus that will favorably distinguish it from the "competitors". I would call this function "replacement". In one window, there is an image in which replacements will be made. This image marks the fragment that needs to be changed by design. An image is placed in another window, and the fragment that will be replaced is marked on it. Suppose we are not satisfied with the hand (face, items of clothing, objects), but there is another image in which the hand (face, items of clothing, objects) is exactly what we would like to receive. We highlight the "bad hand" in one image, highlight the "good hand" on the donor. Instead of taking a long time to select the necessary product, we simply demonstrate the required change. I'm sure the Focus is capable of more.

**Reply by IPv6** [reply]
try ComfyUI, it is more suitable for fine-grained workflow.

**Comment by theaccofai** [comment]
How to fix hands and fingers? I generated a female model that is laying on the sun bed, everything is great, but the fingers on her hand and toes are incorrect (

**Reply by mashb1t** [reply]
use manual inpaint, use the auto masking feature of [my fork](https://github.com/mashb1t/Fooocus) or wait for https://github.com/mashb1t/Fooocus/discussions/42 (also see https://github.com/lllyasviel/Fooocus/discussions/3122).

**Reply by antoniormrzz** [reply]
Inpaint them until you get something close to good, then add some lines with photoshop and try to inpaint again. You can also try to fix things in photoshop.

**Reply by theaccofai** [reply]
Inpaint default, improve detail, or modify content ?

**Reply by theaccofai** [reply]
> Inpaint default, improve detail, or modify content ?

I also tried prompt: (hands,fingers:1,5)

**Reply by antoniormrzz** [reply]
I used default but you can try improve too. try putting negative prompt like deformed hand, deformed fingers, mutated, etc. You can find good examples of negatives on CivitAI.

**Comment by trumantian** [comment]
Anyone knows how to avoid the object editing during changing background?
![image](https://github.com/user-attachments/assets/a3a185e9-15cc-49be-9e8a-3cf40eaaa9b7)

**Comment by ErgoR3AL** [comment]
Hello there could you explain me how to change shapes and colors of the hairs of a woman the best way possible? Thank you

**Reply by Yogesh-DevHub** [reply]
Use inpaint, select hair then select inpaint method as modify content. the modify prompt can be 'red hair bun' and generate.
reference video: https://youtu.be/vRKkj_f9cJk?t=205&si=V2OcLcix_NqMv-9K
 
For changing hair color, you can use the advance masking feature by adding detection prompt 'hair' after getting mask, use modify content with prompt to change the color.
https://youtu.be/4oWGtIHFlCQ?t=517&si=5XMLU79vDHZfEm1Z

**Reply by ErgoR3AL** [reply]
I will try thank you

**Comment by XDM-Inc** [comment]
> Note that this method, in general, expects processing generated images with unchanged or minorly changed prompts. We encourage users to drag image like this:

were you saying we can drag and drop or that you would like to see the feature? because for me dragging and dropping from a result back to a source window does not work.

OS: fedora linux
Browser:firerox

**Reply by ghost** [reply]
It works for me.

**Comment by sagina0t0kiou-coder** [comment]
As of 2025/08/17, lots of images in the document are 404. I'd be appreciated if you fix them
